include/cocaine/trace/trace.hpp

```diff
@@ -0,0 +1,290 @@
+/*
+    Copyright (c) 2011-2015 Anton Matveenko <antmat@me.com>
+    Copyright (c) 2011-2015 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser Gene ral Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#ifndef COCAINE_TRACER_TRACER_HPP
+#define COCAINE_TRACER_TRACER_HPP
+#include <blackhole/scoped_attributes.hpp>
+#include "cocaine/forwards.hpp"
+
+#include <random>
+
+namespace cocaine { namespace tracer {
+
+class attribute_scope_t;
+class span_t;
+class logger_t;
+
+class new_trace_scope_t;
+class trace_reset_scope_t;
+class trace_push_scope_t;
+class trace_restore_scope_t;
+
+template <class F>
+class callable_wrapper_t;
+
+typedef std::shared_ptr<span_t> span_ptr_t;
+
+/*
+ ********************************************
+ ************** Public API ******************
+ ********************************************
+*/
+
+inline
+span_ptr_t
+current_span();
+
+inline
+void
+set_service_name(std::string name);
+
+inline void
+set_logger(std::unique_ptr<logger_t> logger);
+
+inline void
+log(std::string message);
+
+//Used to create a new trace. For example in client application.
+class new_trace_scope_t {
+public:
+    inline
+    new_trace_scope_t(std::string rpc_name);
+
+    inline
+    ~new_trace_scope_t();
+private:
+    std::unique_ptr<attribute_scope_t> attr_scope;
```

Comment:
So okay, let it be destroyed. We've already logged them anyway. All information we need now - are current trace/span/parent, because we're now at the top of the new stack.